lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/session/PropertyMetadataUtil.java

```diff
@@ -68,4 +99,154 @@ public static PropertyMetadata<Duration> durationProperty(String name, String de
                 },
                 Duration::toString);
     }
+
+    public static PropertyBuilder property(String name)
+    {
+        return new PropertyBuilder(name);
+    }
+
+    public static class PropertyBuilder
+    {
+        private final String name;
+        private String description;
+        private Type sqlType;
+        private Class<?> javaType;
+        private @Nullable Object defaultValue;
+        private boolean hidden;
+
+        private PropertyBuilder(String name)
+        {
+            this.name = requireNonNull(name, "name is null");
+        }
+
+        @CanIgnoreReturnValue
+        public <C, T> PropertyBuilder fromConfig(C config, Function<C, T> getterMethodReference)
+        {
+            Class<?> configClass = config.getClass();
+            Method getter = findMethodByReference(configClass, configInstance -> {
+                //noinspection unchecked
+                T ignored = getterMethodReference.apply((C) configInstance);
+            });
+            description = findConfigDescriptionByGetter(configClass, getter)
+                    .orElseThrow(() -> new IllegalArgumentException("Config associated with %s.%s has no description".formatted(configClass.getName(), getter.getName())));
+            // Session properties are never primitives, as they are accessed with generic `ConnectorSession.getProperty(String, Class)`
+            javaType = Primitives.wrap(getter.getReturnType());
+            defaultValue = getterMethodReference.apply(config);
+
+            if (javaType == Boolean.class) {
+                sqlType = BOOLEAN;
+            }
+            if (javaType == Integer.class) {
+                sqlType = INTEGER;
+            }
+            if (javaType == Long.class) {
+                sqlType = BIGINT;
+            }
+            if (javaType == String.class) {
+                sqlType = VARCHAR;
+            }
+
+            return this;
+        }
+
+        @CanIgnoreReturnValue
+        public PropertyBuilder hidden()
+        {
+            hidden = true;
+            return this;
+        }
+
+        public PropertyMetadata<?> build()
+        {
+            Function<Object, Object> decoder = Functions.identity();
+            Function<Object, Object> encoder = Functions.identity();
+
+            //noinspection unchecked,rawtypes
+            return new PropertyMetadata(
+                    name,
+                    description,
+                    sqlType,
+                    javaType,
+                    defaultValue,
+                    hidden,
+                    decoder,
+                    encoder);
+        }
+    }
+
+    private static <C> Method findMethodByReference(Class<C> clazz, Consumer<C> methodReference)
+    {
+        List<Method> calledMethods = new ArrayList<>(1);
+        try (DynamicType.Unloaded<C> interceptingClass = new ByteBuddy()
+                .subclass(clazz)
```

Comment:
This is not going to work in general. Config classes should are not expected to be extensible (in fact, they shouldn't be, and should all be final). Once we migrate the config system to use records and builders (or withers, when they make it to Java one day - https://mail.openjdk.org/pipermail/amber-spec-experts/2022-June/003461.html), this will stop working, too.

I also have concerns about adding bytecode manipulation libraries as dependencies. Those are the kinds of libraries that make it harder to migrate to new versions of Java if they are not diligent about updating to support new class formats.